### DISCUSSION TITLE
Mismatch between images displayed and named in /Resources

### DISCUSSION THREAD

**Discussion by frischzenger** [pymupdf/PyMuPDF#2748] [Q&A]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
When extract image from current page, it will returns all page's images

## To Reproduce (mandatory)
def extract_image():
    doc = fitz.open("xxxxxxx.pdf") # open a document

    for page_index in range(len(doc)): # iterate over pdf pages
        page = doc[page_index] # get the page
        image_list = page.get_images()
		
	    # print the number of images found on the page
        if image_list:
            print(f"Found {len(image_list)} images on page {page_index}")
        else:
            print("No images found on page", page_index)
            
        for image_index, img in enumerate(image_list, start=1): # enumerate the image list
            xref = img[0] # get the XREF of the image
            pix = fitz.Pixmap(doc, xref) # create a Pixmap

            if pix.n - pix.alpha > 3: # CMYK: convert to RGB first
                pix = fitz.Pixmap(fitz.csRGB, pix)

            pix.save(".\\tmp\\page_%s-image_%s.png" % (page_index, image_index)) # save the image as png
            pix = None

## Expected behavior (optional)
When extract image from current page, it will returns current page's images

## Your configuration (mandatory)
 - Window11 home edition
 - Anaconda3 python3.10
 - PyMuPDF version('1.23.5', '1.23.4', '20231011000001')

**Comment by julian-smith-artifex-com** [comment]
Do you have an example PDF that shows this problem?

**Comment by frischzenger** [comment]
> Do you have an example PDF that shows this problem?

https://www.goodix.com/zh/docview/GProgrammer%E7%94%A8%E6%88%B7%E6%89%8B%E5%86%8C?objectId=108&objectType=document&version=503

**Comment by JorjMcKie** [comment]
This is not a bug:
The page's `/Resources` object names all these images, although the page itself does not display them. in PDF it is possible to do that.
It was the decision of the PDF creator to do so - whatever the motives.
If you want to change that, you can execute `page.clean_contents()` and save the PDF again using garbage option 2 or higher.
If you want to know which images the page page actually displays, use page.get_image_info()`.

**Comment by frischzenger** [comment]
> This is not a bug: The page's `/Resources` object names all these images, although the page itself does not display them. in PDF it is possible to do that. It was the decision of the PDF creator to do so - whatever the motives. If you want to change that, you can execute `page.clean_contents()` and save the PDF again using garbage option 2 or higher. If you want to know which images the page page actually displays, use page.get_image_info()`.

when I use page.get_image_info(), and I get:
>[{'number': 1, 'bbox': (148.01400756835938, 53.072998046875, 463.9859924316406, 279.843994140625), 'transform': (315.97198486328125, 0.0, -0.0, 226.77099609375, 148.01400756835938, 53.072998046875), 'width': 503, 'height': 361, 'colorspace': 1, 'cs-name': 'Indexed(255,DeviceRGB)', 'xres': 96, 'yres': 96, 'bpc': 8, 'size': 65723}, {'number': 6, 'bbox': (148.01400756835938, 397.43402099609375, 463.9859924316406, 624.2050170898438), 'transform': (315.97198486328125, 0.0, -0.0, 226.77099609375, 148.01400756835938, 397.43402099609375), 'width': 503, 'height': 361, 'colorspace': 1, 'cs-name': 'Indexed(255,DeviceRGB)', 'xres': 96, 'yres': 96, 'bpc': 8, 'size': 64160}]

```
page = doc[page_index] # get the page
image_list = page.get_images()
#which image is :
im_idx = page.get_image_info()[0]['number']
image_list[im_idx ]
```

and I show the image by the field 'number', but it's not correspondes to the image of that page displayed

**Comment by JorjMcKie** [comment]
As the documentation says:
There is no reliable relationship between images referenced in `page.get_images()` and the images actually displayed.
You can have 500 images  listed in `page.get_images()` and the page shows no image at all.
And the reverse is also possible: the page may display an image which is not in that other list.

**Comment by JorjMcKie** [comment]
In addition, an image occuring with some number in `page.get_image_info()` does not mean that image is shown as the first on the page.
That number only means the sequence number of the display command. The image itself may still be shown at the bottom of the page.
